packages/xod-project/src/expandVariadicNodes.js

```diff
@@ -0,0 +1,344 @@
+import { getExpandedPatchPath } from './patchPathUtils';
+
+//
+// get dependencies
+//
+const getDependenciesMap = (project, patchPaths, depsMap) => {
+  if (R.isEmpty(patchPaths)) {
+    return depsMap;
+  }
+
+  const [currentPatchPath, ...restPatchPaths] = patchPaths;
+
+  if (R.has(currentPatchPath, depsMap)) {
+    return getDependenciesMap(project, restPatchPaths, depsMap);
+  }
+
+  const currentPatchDeps = R.compose(
+    R.map(Node.getNodeType),
+    Patch.listNodes,
+    Project.getPatchByPathUnsafe(currentPatchPath),
+  )(project);
+
+  return getDependenciesMap(
+    project,
+    R.concat(restPatchPaths, currentPatchDeps),
+    R.assoc(currentPatchPath, currentPatchDeps, depsMap)
+  );
+};
+
+// :: PatchPath -> Project -> [PatchPath]
+export const getPatchDependencies = R.curry(
+  (entryPatchPath, project) => R.compose(
+    R.uniq,
+    R.unnest,
+    R.values,
+    getDependenciesMap
+  )(project, [entryPatchPath], {})
+);
+
+//
+// expanding a single patch
+//
+
+const getNodeX = R.pipe(Node.getNodePosition, R.prop('x'));
+
+const nodeIdLens = R.lens(Node.getNodeId, R.assoc('id'));
+
+
+// helpers for creating nodes inside expanded patch
+
+const createAdditionalValueTerminalGroups = (
+  patch,
+  desiredArityLevel,
+  originalTerminalNodes,
+  variadicPins,
+) => {
+  const arityStep = R.compose(
+    explodeMaybe('Patch is guaranteed to be variadic at this point'),
+    Patch.getArityStepFromPatch
+  )(patch);
+
+  // for positioning new terminals
+  const rightmostInputTerminalX = R.compose(
+    xs => Math.max(...xs),
+    R.map(getNodeX),
+    R.filter(Node.isInputPinNode)
+  )(originalTerminalNodes);
+
+
+  // :: [ Node ]
+  const originalValueTerminalNodes = R.compose(
+    R.sortBy(getNodeX),
+    R.map(R.compose(
+      R.flip(Patch.getNodeByIdUnsafe)(patch),
+      Pin.getPinKey,
+    )),
+    R.prop('value'),
+  )(variadicPins);
+
+  const DISTANCE_BETWEEN_TERMINALS = 50;
+  const valueTerminalsGroupWidth = arityStep * DISTANCE_BETWEEN_TERMINALS;
+  const getValueTerminalX = (terminalGroupIndex, terminalIndex) => (
+    rightmostInputTerminalX +
+    DISTANCE_BETWEEN_TERMINALS +
+    // because terminalGroupIndex always starts from 2
+    ((terminalGroupIndex - 2) * valueTerminalsGroupWidth) +
+    (terminalIndex * DISTANCE_BETWEEN_TERMINALS)
+  );
+
+  // :: [ [Node] ]
+  return R.compose(
+    R.map(terminalGroupIndex => R.addIndex(R.map)((node, index) =>
+      R.compose(
+        R.over(
+          nodeIdLens,
+          // TODO: extract to a public util. where to put it?
+          nodeId => `${nodeId}-$${terminalGroupIndex}`
+        ),
+        Node.setNodePosition({
+          x: getValueTerminalX(terminalGroupIndex, index),
+          y: 0,
+        })
+      )(node),
+      originalValueTerminalNodes
+    )),
+    R.range(2), // [2; desiredArityLevel]
+    R.inc
+  )(desiredArityLevel);
+};
+
+// TODO: better name?
+const createExpandedNodeInstances = (patch, desiredArityLevel) => {
```

Comment:
`createExpansionNodes`? Because a node is already === an instance (of the patch).